### DISCUSSION TITLE
General Feedback

### DISCUSSION THREAD

**Discussion by tannermeade** [tannermeade/data-migrator#2] [General]
I'd love your feedback as you use Data Migrator!

- What are you wanting to use it for?
- What type of data sources are you wanting to connect to?
- What problems are you having?
- Any comments and feedback are great! Thanks!

**Comment by gepd** [comment]
Hey @tannermeade first to all, thank you for work on this, and open source it, I'll just give a try and test it in my local machine. 
I was wondering, do you have any plans to support something like migration from local server to production server and viceversa?

Currently this task are very time consuming in my workflow and I'm looking a way to improve that

**Reply by tannermeade** [reply]
Thank you for the feedback! The kind of help you've been giving to this project is very valuable and useful! 😁

You are totally right that it is hard to deploy data from a local/dev Appwrite instance to a production environment.

The tool will eventually be able to have Appwrite as a data source (not just a destination). I've been deciding on how exactly I want to implement that, however it will probably involve:
- importing a collection schema into the tool
- using a cloud function to aggregate the data into storage zip files
- each cloud execution will be responsible for aggregating a portion of the data
- once the aggregated data file is downloaded and in the conversion pipeline, it will be deleted from the source Appwrite server
- this will be significantly simpler once v0.13 of Appwrite comes out with sync cloud functions.

**Reply by gepd** [reply]
Thanks for your answer

Yes there are some details to check, for example sometimes you not only need to update the data but the collection as well. Many times I have new fields or collection in my local instance and I have to commit that changes to the production server, so would be good to have this feature where you can sync data, fields and collection.

I'm waiting for v.0.13 

Now that you open sourced the code, I'll be happy to help with this project!

**Reply by tannermeade** [reply]
> Now that you open sourced the code, I'll be happy to help with this project!

Thank you so much for the help!

Right now, when a collection is imported it is marked as immutable (the SchemaMap.mutable == false) which defines if the UI will be read only.

This is mostly just to have continuity with the experience because there isn't any functionality yet to resolve differences between the tool's schema model and the remote schema (unless it doesn't exist, then it'll create the collection).

If you were thinking about helping to design or implement this feature feel free to open a new discussion and start putting the feature together conceptually. I can answer any questions you have about the current architecture and when we get it fleshed out we can move forward with implementing it 👍

**Reply by tannermeade** [reply]
> Yes there are some details to check, for example sometimes you not only need to update the data but the collection as well. Many times I have new fields or collection in my local instance and I have to commit that changes to the production server, so would be good to have this feature where you can sync data, fields and collection.

@gepd Today I did a lot of work on resolving conflicts between a collection in the tool and the corresponding collection in Appwrite.

When you start the conversion, it'll cross-reference all the collections in the tool (that have an id) with the ones in Appwrite, and provide a series of dialogs that will tell you exactly what change needs to happen and if you'd like that specific change to be made. This applies to collection metadata like the collection name and attributes (type, default, size, etc).

If you tap confirm on all of them, it'll make the changes and will start the conversion process.

Still some work and testing to do, but got a big bulk of work done today.

Also I've made some changes so you can undo-redo schema changes to the source & destination. 🚀

**Reply by gepd** [reply]
That looks like very good updates, I'm looking foward to test that changes.

I was thinking about the collection sync between appwrite instances the other day, and it may be like something separte from the current window you have. I have noted that you have a button at the beggining of app where you enter to the migrator, you can add more button for specific functions.

For example now you have  CSV -> Appwrite but would be good to have 

Appwrite -> CSV
Appwrite -> Appwrite (Collection)
Appwrite -> Appwrite (Data)
Firebase -> Appwrite

And so on. That could help to have a tool more user friendly as you won't have so many options in the same interface, and good for maintain it as well

**Reply by tannermeade** [reply]
> That looks like very good updates, I'm looking foward to test that changes.

@gepd 
I just pushed the initial update for this feature (updating collections/attributes). Feel free to test away as there are definitely bugs (main use case work well of course). I won't have as much time for the next while compared to the past month or 2 (when I've built this and got it open-sourced), but I'm hoping to get some more important features out and maybe get some good code coverage on the important core parts so people can feel some integrity & reliability going forward.

> I was thinking about the collection sync between appwrite instances the other day, and it may be like something separte from the current window you have. I have noted that you have a button at the beggining of app where you enter to the migrator, you can add more button for specific functions.

I'm thinking of making the titles "CSV" and "Appwrite" basically dropdown menus where you can choose from a list of available data origins. Right now the CSV DataOrigin only is good as a source and the Appwrite only as a destination. So it hasn't made sense to focus on that quite yet, however that is the direction I'm thinking of going.
